File: packages/desktop/src/core/sync-client.ts
/**
 * Client-side view of the sync server. Abstracted behind a Transport so the app
 * core can be tested against the in-process VaultService OR a real HTTP server
 * with identical behavior.
 */

export interface RemoteItem {
  id: string;
  ciphertext: string;
  version: number;
  updatedAt: string;
}

export type PushOutcome =
  | { status: "ok"; version: number }
  | { status: "conflict"; serverVersion: number; server: RemoteItem | null };

/** The minimal contract the desktop app needs from the server. */
export interface Transport {
  pull(since?: string): Promise<RemoteItem[]>;
  push(id: string, ciphertext: string, baseVersion: number | null): Promise<PushOutcome>;
}

/** HTTP transport for production (Bearer token auth). */
export class HttpTransport implements Transport {
  constructor(
    private readonly baseUrl: string,
    private readonly token: string,
    private readonly fetchImpl: typeof fetch = fetch
  ) {}

  private headers() {
    return {
      authorization: `Bearer ${this.token}`,
      "content-type": "application/json",
    };
  }

  async pull(since?: string): Promise<RemoteItem[]> {
    const u = new URL(`${this.baseUrl}/vault/items`);
    if (since) u.searchParams.set("since", since);
    const res = await this.fetchImpl(u, { headers: this.headers() });
    if (!res.ok) throw new Error(`pull failed: ${res.status}`);
    return (await res.json()).items as RemoteItem[];
  }

  async push(id: string, ciphertext: string, baseVersion: number | null): Promise<PushOutcome> {
    const res = await this.fetchImpl(`${this.baseUrl}/vault/items/${id}`, {
      method: "PUT",
      headers: this.headers(),
      body: JSON.stringify({ ciphertext, baseVersion }),
    });
    return (await res.json()) as PushOutcome;
  }
}

File: packages/mobile/lib/core/sync_client.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

/// Client-side view of the sync server — the same Transport contract the
/// desktop core uses, so the mobile app talks to the identical API.
class RemoteItem {
  final String id;
  final String ciphertext;
  final int version;
  final String updatedAt;
  RemoteItem({
    required this.id,
    required this.ciphertext,
    required this.version,
    required this.updatedAt,
  });
  factory RemoteItem.fromJson(Map<String, dynamic> j) => RemoteItem(
        id: j['id'] as String,
        ciphertext: j['ciphertext'] as String,
        version: j['version'] as int,
        updatedAt: (j['updatedAt'] ?? '') as String,
      );
}

class PushOutcome {
  final String status; // "ok" | "conflict"
  final int? version;
  final RemoteItem? server;
  PushOutcome({required this.status, this.version, this.server});
  bool get ok => status == 'ok';
  factory PushOutcome.fromJson(Map<String, dynamic> j) => PushOutcome(
        status: j['status'] as String,
        version: j['version'] as int?,
        server: j['server'] == null
            ? null
            : RemoteItem.fromJson(j['server'] as Map<String, dynamic>),
      );
}

abstract class Transport {
  Future<List<RemoteItem>> pull({String? since});
  Future<PushOutcome> push(String id, String ciphertext, int? baseVersion);
}

/// HTTP transport for production (Bearer token auth).
class HttpTransport implements Transport {
  final String baseUrl;
  final String token;
  final http.Client _client;
  HttpTransport(this.baseUrl, this.token, {http.Client? client})
      : _client = client ?? http.Client();

  Map<String, String> get _headers => {
        'authorization': 'Bearer $token',
        'content-type': 'application/json',
      };

  @override
  Future<List<RemoteItem>> pull({String? since}) async {
    final uri = Uri.parse('$baseUrl/vault/items')
        .replace(queryParameters: since != null ? {'since': since} : null);
    final res = await _client.get(uri, headers: _headers);
    if (res.statusCode != 200) {
      throw Exception('pull failed: ${res.statusCode}');
    }
    final items = (jsonDecode(res.body)['items'] as List<dynamic>);
    return items
        .map((e) => RemoteItem.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  @override
  Future<PushOutcome> push(String id, String ciphertext, int? baseVersion) async {
    final res = await _client.put(
      Uri.parse('$baseUrl/vault/items/$id'),
      headers: _headers,
      body: jsonEncode({'ciphertext': ciphertext, 'baseVersion': baseVersion}),
    );
    return PushOutcome.fromJson(jsonDecode(res.body) as Map<String, dynamic>);
  }
}
